In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/vineetasinha/Documents/dataanalytics/tasks/Data-Immersion/Achievement4/Instacart_Basket_Analysis'

In [3]:
path

'/Users/vineetasinha/Documents/dataanalytics/tasks/Data-Immersion/Achievement4/Instacart_Basket_Analysis'

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02_Data','Original_Data','orders.csv'),index_col=False)

In [5]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [6]:
df_prods = pd.read_csv(os.path.join(path,'02_Data','Original_Data','products.csv'),index_col=False)

In [7]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [8]:
df_ords.drop(column = ['eval_set'])

TypeError: DataFrame.drop() got an unexpected keyword argument 'column'

In [9]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [10]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [11]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [12]:
df_ords.rename(columns = {'order_dow':'order_day_of_week'},inplace = True)

In [13]:
df_ords.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [14]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [15]:
df_ords[order_id].dtype

NameError: name 'order_id' is not defined

In [16]:
df_ords['order_id'].dtype

dtype('O')

In [17]:
df_dep = pd.read_csv(os.path.join(path, '02_Data','Original_Data','departments.csv'),index_col=False)

In [18]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [19]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [20]:
df_dep_t = df_dep.T

In [21]:
df_dep_t.head()

0
department_id  department
1                  frozen
2                   other
3                  bakery
4                 produce

In [22]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [24]:
new_header = df_dep_t.iloc[0]

In [25]:
new_header

0    department
Name: department_id, dtype: object

In [26]:
df_dep_t_new = df_dep_t[1:]

In [27]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [28]:
df_dep_t_new.columns = new.header

NameError: name 'new' is not defined

In [29]:
df_dep_t_new.columns = new_header

In [30]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [31]:
data_dict = df_dep_t_new.to_dict('index')

In [32]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [33]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [34]:
print(data_dict.get(19))

None


In [35]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [36]:
df_snacks = df_prods[df_prods['department_id']==19]

In [37]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# New Section - Exercise 4.4


In [38]:
df_ords.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0  2539329        1    prior             1                  2   
1  2398795        1    prior             2                  3   
2   473747        1    prior             3                  3   
3  2254736        1    prior             4                  4   
4   431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [39]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [40]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329        1             1                  2                  8   
1  2398795        1             2                  3                  7   
2   473747        1             3                  3                 12   
3  2254736        1             4                  4                  7   
4   431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Answer - 2

In [41]:
df_ords.describe()

user_id  order_number  order_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06       3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01       2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01       2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00       0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00       1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01       3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01       5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02       6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [42]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [43]:
df_ords['user_id'].dtypes

dtype('O')

In [44]:
df_ords.describe()

order_number  order_day_of_week  order_hour_of_day  \
count  3.421083e+06       3.421083e+06       3.421083e+06   
mean   1.715486e+01       2.776219e+00       1.345202e+01   
std    1.773316e+01       2.046829e+00       4.226088e+00   
min    1.000000e+00       0.000000e+00       0.000000e+00   
25%    5.000000e+00       1.000000e+00       1.000000e+01   
50%    1.100000e+01       3.000000e+00       1.300000e+01   
75%    2.300000e+01       5.000000e+00       1.600000e+01   
max    1.000000e+02       6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

# Answer - 3

In [45]:
df_ords.rename(columns = {'order_hour_of_day': 'order_hour'},inplace = True)

In [46]:
df_ords.describe()

order_number  order_day_of_week    order_hour  days_since_prior_order
count  3.421083e+06       3.421083e+06  3.421083e+06            3.214874e+06
mean   1.715486e+01       2.776219e+00  1.345202e+01            1.111484e+01
std    1.773316e+01       2.046829e+00  4.226088e+00            9.206737e+00
min    1.000000e+00       0.000000e+00  0.000000e+00            0.000000e+00
25%    5.000000e+00       1.000000e+00  1.000000e+01            4.000000e+00
50%    1.100000e+01       3.000000e+00  1.300000e+01            7.000000e+00
75%    2.300000e+01       5.000000e+00  1.600000e+01            1.500000e+01
max    1.000000e+02       6.000000e+00  2.300000e+01            3.000000e+01

# Answer - 4

In [47]:
df_ords

order_id user_id  order_number  order_day_of_week  order_hour  \
0        2539329       1             1                  2           8   
1        2398795       1             2                  3           7   
2         473747       1             3                  3          12   
3        2254736       1             4                  4           7   
4         431534       1             5                  4          15   
...          ...     ...           ...                ...         ...   
3421078  2266710  206209            10                  5          18   
3421079  1854736  206209            11                  4          10   
3421080   626363  206209            12                  1          12   
3421081  2977660  206209            13                  1          12   
3421082   272231  206209            14                  6          14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [48]:
df_ords['order_hour'].value_counts(dropna = False)

order_hour
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

10am seems to be the busiest hour for placing order

# Answer - 5 

In [50]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [52]:
df_prods[df_prods['department_id']==4]

product_id                                      product_name  aisle_id  \
30             31                                White Pearl Onions       123   
42             43                               Organic Clementines       123   
44             45                                 European Cucumber        83   
65             66                         European Style Spring Mix       123   
88             89                    Yogurt Fruit Dip Sliced Apples       123   
...           ...                                               ...       ...   
49582       49578                                Black Garlic Bulbs       123   
49623       49619                                        Opo Squash        83   
49639       49635  Baby Food Blueberry, Parsnip & Buckwheat Stage 2        83   
49661       49657                                 Cabernet Tomatoes        83   
49687       49683                                    Cucumber Kirby        83   

       department_id  prices  
30                 4     7.5  
42                 4    11.5  
44                 4    14.3  
65                 4    11.7  
88                 4    12.6  
...              ...     ...  
49582              4     8.0  
49623              4    12.7  
49639              4    12.5  
49661              4     8.3  
49687              4    13.2  

[1684 rows x 5 columns]

In [53]:
print(data_dict.get('4'))

{'department': 'produce'}


The value of 4 means 'produce' in df_prod dataframe

# Answer - 6 

In [54]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [55]:
df_breakfast = df_prods[df_prods['department_id']==14]

In [56]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# Answer - 7 

In [57]:
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [58]:
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Answer -8 - Total row counts in the dataframes is 7650.

# Answer - 9  Extract all the information you can about  user_id = 1

In [59]:
df_ords

order_id user_id  order_number  order_day_of_week  order_hour  \
0        2539329       1             1                  2           8   
1        2398795       1             2                  3           7   
2         473747       1             3                  3          12   
3        2254736       1             4                  4           7   
4         431534       1             5                  4          15   
...          ...     ...           ...                ...         ...   
3421078  2266710  206209            10                  5          18   
3421079  1854736  206209            11                  4          10   
3421080   626363  206209            12                  1          12   
3421081  2977660  206209            13                  1          12   
3421082   272231  206209            14                  6          14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [60]:
df_info = df_ords.loc[df_ords['user_id'].isin([1])]

In [61]:
df_info

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour, days_since_prior_order]
Index: []

In [62]:
df_info = df_ords[df_ords['user_id']==1]

In [63]:
df_info

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour, days_since_prior_order]
Index: []

In [64]:
df_info.head()

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour, days_since_prior_order]
Index: []

In [65]:
df_info = df_ords[df_ords['user_id']=='1']

In [66]:
df_info.head()

order_id user_id  order_number  order_day_of_week  order_hour  \
0  2539329       1             1                  2           8   
1  2398795       1             2                  3           7   
2   473747       1             3                  3          12   
3  2254736       1             4                  4           7   
4   431534       1             5                  4          15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Answer - 10 - Basic Stat about user_id 1

In [67]:
df_info.describe()

order_number  order_day_of_week  order_hour  days_since_prior_order
count     11.000000          11.000000   11.000000               10.000000
mean       6.000000           2.636364   10.090909               19.000000
std        3.316625           1.286291    3.477198                9.030811
min        1.000000           1.000000    7.000000                0.000000
25%        3.500000           1.500000    7.500000               14.250000
50%        6.000000           3.000000    8.000000               19.500000
75%        8.500000           4.000000   13.000000               26.250000
max       11.000000           4.000000   16.000000               30.000000

In [68]:
df_ords.to_csv(os.path.join(path, '02_Data','Prepared_Data','orders_wrangled.csv'))

In [69]:
df_dep_t_new.to_csv(os.path.join(path, '02_Data','Prepared_Data','departments_wrangled.csv'))